drivers/crc/crc_stm32.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) 2024 Brill Power Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(crc_stm32, CONFIG_CRC_HW_LOG_LEVEL);
+
+#include <errno.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/drivers/crc.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <soc.h>
+#include <stm32_ll_crc.h>
+
+#define DT_DRV_COMPAT st_stm32_crc
+
+struct crc_stm32_cfg {
+	CRC_TypeDef *base;
+	struct stm32_pclken pclken;
+};
+
+struct crc_stm32_data {
+	struct k_sem sem;
+};
+
+static void crc_lock(const struct device *dev)
+{
+	struct crc_stm32_data *data = (struct crc_stm32_data *)dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+}
+
+static void crc_unlock(const struct device *dev)
+{
+	struct crc_stm32_data *data = (struct crc_stm32_data *)dev->data;
+
+	k_sem_give(&data->sem);
+}
+
+static int crc_stm32_begin(const struct device *dev, struct crc_ctx *ctx)
+{
+	/* Ensure ctx is not currently being updated */
+	if (ctx->state == CRC_STATE_IN_PROGRESS) {
+		return -EBUSY;
+	}
```

Comment:
I think you can have a race issue here, checking the state outside of the lock, but changing it inside the lock.